server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/http/SessionRoutes.scala

```diff
@@ -0,0 +1,71 @@
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ * *
+ * http://www.apache.org/licenses/LICENSE-2.0                 *
+ * *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ * ***************************************************************/
+
+package org.apache.james.jmap.http
+
+import java.util.function.BiFunction
+
+import io.netty.handler.codec.http.HttpHeaderNames.CONTENT_TYPE
+import io.netty.handler.codec.http.HttpResponseStatus.OK
+import javax.inject.Inject
+import org.apache.james.jmap.HttpConstants.JSON_CONTENT_TYPE_UTF8
+import org.apache.james.jmap.exceptions.UnauthorizedException
+import org.apache.james.jmap.http.SessionRoutes.JMAP_SESSION
+import org.apache.james.jmap.model.Session
+import org.apache.james.jmap.{JMAPRoutes, Serializer}
+import org.reactivestreams.Publisher
+import org.slf4j.{Logger, LoggerFactory}
+import reactor.core.scala.publisher.SMono
+import reactor.core.scheduler.Schedulers
+import reactor.netty.http.server.{HttpServerRequest, HttpServerResponse, HttpServerRoutes}
+
+object SessionRoutes {
+  private val JMAP_SESSION = "/jmap/session"
+  private val LOGGER = LoggerFactory.getLogger(classOf[SessionRoutes])
+}
+
+@Inject
+class SessionRoutes(val authFilter: AuthenticationReactiveFilter,
+                    val serializer: Serializer = new Serializer) extends JMAPRoutes {
+
+  val logger: Logger = SessionRoutes.LOGGER
+  val generateSession: BiFunction[HttpServerRequest, HttpServerResponse, Publisher[Void]] =
+    (request, response) => SMono.fromPublisher(authFilter.authenticate(request))
+      .map(_.getUser)
+      .map(SessionSupplier.generate)
+      .flatMap(session => sendRespond(session, response))
+      .onErrorResume(throwable => errorHandling(throwable, response))
+      .subscribeOn(Schedulers.elastic())
+
+  override def define(builder: HttpServerRoutes): HttpServerRoutes = {
+    builder.get(JMAP_SESSION, generateSession)
+  }
+
+  private def sendRespond(session: Session, resp: HttpServerResponse): SMono[Void] =
+    SMono.fromPublisher(resp.header(CONTENT_TYPE, JSON_CONTENT_TYPE_UTF8)
+      .status(OK)
+      .sendString(SMono.fromCallable(() => serializer.serialize(session)))
+      .`then`())
+
+  def errorHandling(throwable: Throwable, response: HttpServerResponse): SMono[Void] =
+    throwable match {
+      case _: UnauthorizedException => SMono.fromPublisher(handleAuthenticationFailure(response, throwable))
```

Comment:
you could have `fromPublisher` around the pattern matching (or even outside the method)